diff --git a/gcc/vhdl/Make-lang.in b/gcc/vhdl/Make-lang.in
index e9c9b23..4fee680 100644
--- a/gcc/vhdl/Make-lang.in
+++ b/gcc/vhdl/Make-lang.in
@@ -120,7 +120,7 @@ AGCC_OBJS := $(AGCC_LOCAL_OBJS) \
 ortho-lang.o: $(agcc_srcdir)/ortho-lang.c \
  $(AGCC_GCCOBJ_DIR)gcc/gtype-vhdl.h \
  $(AGCC_GCCOBJ_DIR)gcc/gt-vhdl-ortho-lang.h
-	$(CC) -c -o $@ $< $(AGCC_CFLAGS)
+	$(CC) -c $(INCLUDES) -o $@ $< $(AGCC_CFLAGS)
 
 agcc-clean: force
 	$(RM) -f $(agcc_objdir)/*.o
